src/subject.rs

```diff
@@ -133,19 +135,32 @@ async fn get_subject_evaluation_projects(
     let evaluation_projects: Vec<EvaluationProject> =
         serde_json::from_value(response["data"]["evaluationProjectList"].clone())
             .expect("Failed to get evaluation projects");
+    let total_proportion: f64 = evaluation_projects
+        .iter()
+        .filter(|evaluation_project| !evaluation_project.score_is_null)
+        .map(|evaluation_project| evaluation_project.proportion)
+        .sum();
     evaluation_projects
+        .into_iter()
+        .filter(|evaluation_project| !evaluation_project.score_is_null)
+        .map(|mut evaluation_project| {
+            evaluation_project.adjusted_proportion = Some(evaluation_project.proportion / total_proportion * 100.0);
+            evaluation_project
+        })
+        .collect()
 }
 
 fn get_subject_score(evaluation_projects: &[EvaluationProject]) -> f64 {
-    let mut total_score = 0.0;
-    let mut total_proportion = 0.0;
-    for evaluation_project in evaluation_projects {
-        total_score += evaluation_project.proportion * evaluation_project.score;
-        if !evaluation_project.score_is_null {
-            total_proportion += evaluation_project.proportion;
-        }
-    }
-
+    let total_proportion: f64 = evaluation_projects
+        .iter()
+        .filter(|e| !e.score_is_null)
+        .map(|e| e.proportion)
+        .sum();
+    let total_score: f64 = evaluation_projects
+        .iter()
+        .filter(|e| !e.score_is_null)
+        .map(|e| e.score * e.proportion)
+        .sum();
     total_score / total_proportion
```

Comment:
```suggestion
    evaluation_projects
        .iter()
        .map(|e| e.score * e.adjusted_proportion)
        .sum()
```
This is what I mean.

Reply:
I filtered out null score evaluation subjects, if it should be kept, there's a problem.
The old method will return a NAN when total_proportion is 0, which makes it can be filtered out after when calculating GPA and displaying grades, if we use adjusted_proportion and make adjusted_proportion equal to 0 at the same time, the result will be 0 when total_proportion is 0, which cause subjects that have no valid evaluation projects been calculated as valid subject with a score 0.

We either
1. Filter out null evaluation projects, which is what the function get_subject_evaluation_projects is doing currently.
2. Set adjusted_proportion to Some(NAN) when it's null.
3. Write check in get_subject_score so it will return NAN when no valid evaluation project exist.

If we choose 1, what's in adjusted_proportion doesn't matter, as the null evaluation project is dropped.
If we choose 3, I would prefer the old method more, as it's more clean.